### ISSUE TITLE
Union enum initialized by tuple causing compiler error

### ISSUE THREAD

**Issue by Sirius902** [ziglang/zig#17631]
### Zig Version

0.12.0-dev.1127+32bc07767

### Steps to Reproduce and Observed Behavior

* Use the below `src/main.zig`.
```zig
const Mango = struct {
    pub fn eat(self: Mango) void {
        _ = self;
    }
};

const Fruit = union(enum) {
    mango: Mango,
};

pub fn main() void {
    const fruit: Fruit = .{ .mango = .{} };
    fruit.mango.eat();
}

test "Mango eat" {
    const fruit: Fruit = .{ .mango = .{} };
    fruit.mango.eat();
}
```
* Running `zig build run` will yield the following error message.
```
zig build-exe test-zig Debug native: error: the following command terminated unexpectedly:
/home/blah/dev/zig/zig/build/stage3/bin/zig build-exe /home/blah/dev/zig/test-zig/src/main.zig --cache-dir /home/blah/dev/zig/test-zig/zig-cache --global-cache-dir /home/blah/.cache/zig --name test-zig --listen=-
Build Summary: 0/5 steps succeeded; 1 failed (disable with --summary none)
run transitive failure
└─ run test-zig transitive failure
   ├─ zig build-exe test-zig Debug native failure
   └─ install transitive failure
      └─ install test-zig transitive failure
         └─ zig build-exe test-zig Debug native (reused)
error: the following build command failed with exit code 1:
/home/blah/dev/zig/test-zig/zig-cache/o/096dde1ca0d4af9223dd2716ae2ef386/build /home/blah/dev/zig/zig/build/stage3/bin/zig /home/blah/dev/zig/test-zig /home/blah/dev/zig/test-zig/zig-cache /home/blah/.cache/zig run
```
* Running `zig test src/main.zig` will yield a different but still mysterious compile error.
```
src/main.zig:18:10: error: access of union field 'mango' while field 'mango' is active
    fruit.mango.eat();
    ~~~~~^~~~~~
src/main.zig:7:15: note: union declared here
const Fruit = union(enum) {
              ^~~~~
```
* Now, replace all instances of  `const fruit: Fruit = .{ .mango = .{} };` with `const fruit = Fruit{ .mango = .{} };`.
* `zig build run` will now succeed and `zig test src/main.zig` will succeed with the message `All 1 tests passed.`.

### Expected Behavior

I would expect initializing `fruit` using `const fruit: Fruit = .{ .mango = .{} };` to have the same output from `zig build run` and `zig test src/main.zig` as using `const fruit = Fruit{ .mango = .{} };`.

**Comment by Vexu** [comment]
There is an assertion failure with a debug build:
```
thread 68305 panic: reached unreachable code
Analyzing a.zig: a.zig:main
      %26 = dbg_block_begin()
      %27 = dbg_stmt(2, 5)
      %28 = block_comptime({
        %29 = decl_val("Fruit") token_offset:12:18 to :12:23
        %30 = break(%28, %29)
      }) node_offset:12:18 to :12:23
      %31 = alloc(%28) node_offset:12:5 to :12:43
      %32 = opt_eu_base_ptr_init(%31) node_offset:12:26 to :12:43
      %33 = struct_init_field_ptr(%32, "mango") node_offset:12:38 to :12:41
      %34 = typeof(%33) node_offset:12:38 to :12:41
      %35 = elem_type(%34) node_offset:12:38 to :12:41
      %36 = struct_init_empty_result(%35) node_offset:12:38 to :12:41
      %37 = store_node(%33, %36) node_offset:12:38 to :12:41
      %38 = validate_ptr_struct_init({
        %33 = struct_init_field_ptr(%32, "mango") node_offset:12:38 to :12:41
      }) node_offset:12:26 to :12:43
    > %39 = make_ptr_const(%31) node_offset:12:5 to :12:43
      %40 = dbg_var_ptr(%39, "fruit")
      %41 = dbg_stmt(3, 5)
      %42 = dbg_stmt(3, 10)
      %43 = field_ptr(%39, "mango") node_offset:13:5 to :13:16
      %44 = dbg_stmt(3, 16)
      %45 = dbg_stmt(3, 20)
      %46 = field_call(nodiscard .auto, %43, "eat", []) node_offset:13:5 to :13:22
      %47 = dbg_block_end()
      %48 = restore_err_ret_index(%25, %4294967211)
      %49 = break(%25, @InternPool.Index.void_value)
    For full context, use the command
      zig ast-check -t a.zig

  in a.zig: a.zig:main
    > %25 = block({%26..%49}) node_offset:11:20 to :11:20

/home/vexu/Documents/zig/zig/lib/std/debug.zig:342:14: 0x8d7108c in assert (zig)
    if (!ok) unreachable; // assertion failure
             ^
/home/vexu/Documents/zig/zig/src/Sema.zig:3723:15: 0x9c480ce in resolveComptimeKnownAllocValue (zig)
        assert(mod.intern_pool.typeOf(val) == elem_ty.toIntern());
              ^
/home/vexu/Documents/zig/zig/src/Sema.zig:3659:48: 0x972bae7 in zirMakePtrConst (zig)
    if (try sema.resolveComptimeKnownAllocValue(block, alloc, null)) |val| {
                                               ^
/home/vexu/Documents/zig/zig/src/Sema.zig:1010:70: 0x93db567 in analyzeBodyInner (zig)
            .make_ptr_const               => try sema.zirMakePtrConst(block, inst),
                                                                     ^
```

**Comment by schmee** [comment]
Test passes as of `0.12.0-dev.2760+6fe90a913`.

**Comment by Vexu** [comment]
Fixed by #18736
